src/chai/ArrayManager.cpp

```diff
@@ -264,17 +267,13 @@ void ArrayManager::free(PointerRecord* pointer_record, ExecutionSpace spaceToFre
   for (int space = CPU; space < NUM_EXECUTION_SPACES; ++space) {
     if (space == spaceToFree || spaceToFree == NONE) {
       if (pointer_record->m_pointers[space]) {
+        void* space_ptr = pointer_record->m_pointers[space];
         if (pointer_record->m_owned[space]) {
-          void* space_ptr = pointer_record->m_pointers[space];
 #if defined(CHAI_ENABLE_UM)
           if (space_ptr == pointer_record->m_pointers[UM]) {
             callback(pointer_record,
                      ACTION_FREE,
                      ExecutionSpace(UM));
-            {
-              std::lock_guard<std::mutex> lock(m_mutex);
-              m_pointer_map.erase(space_ptr);
-            }
 
             auto alloc = m_resource_manager.getAllocator(
                 pointer_record->m_allocators[space]);
```

Comment:
The inner loop a couple of lines down seems unnecessary since we already know the space.

Reply:
this is in UM code - UM pointers are the same in all spaces, so you need to 0 them all out.

Reply:
Ah that makes more sense.